napari/utils/progress.py

```diff
@@ -0,0 +1,315 @@
+import inspect
+from typing import Iterable, Optional
+
+from qtpy import QtCore
+from qtpy.QtWidgets import (
+    QApplication,
+    QHBoxLayout,
+    QLabel,
+    QProgressBar,
+    QWidget,
+)
+from tqdm import tqdm
+
+from .._qt.utils import get_viewer_instance
+
+
+def get_pbar(viewer_instance, **kwargs):
+    """Adds ProgressBar to viewer Activity Dock and returns it.
+
+    Parameters
+    ----------
+    viewer_instance : qtViewer
+        current napari qtViewer instance
+
+    Returns
+    -------
+    ProgressBar
+        progress bar to associate with current iterable
+    """
+    pbar = ProgressBar(**kwargs)
+    viewer_instance.activityDock.widget().layout.addWidget(pbar)
+
+    return pbar
+
+
+def get_calling_function_name(max_depth: int):
+    """Inspect stack up to max_depth and return first function name outside of progress.py"""
+    for finfo in inspect.stack()[2:max_depth]:
+        if not finfo.filename.endswith("progress.py"):
+            return finfo.function
+
+    return None
+
+
+_tqdm_kwargs = {
+    p.name
+    for p in inspect.signature(tqdm.__init__).parameters.values()
+    if p.kind is not inspect.Parameter.VAR_KEYWORD and p.name != "self"
+}
+
+
+class progress(tqdm):
+    """This class inherits from tqdm and provides an interface for
+    progress bars in the napari viewer. Progress bars can be created
+    directly by wrapping an iterable or by providing a total number
+    of expected updates.
+
+    See tqdm.tqdm API for valid args and kwargs: https://tqdm.github.io/docs/tqdm/
+
+    Also, any keyword arguments to the :class:`ProgressBar` `QWidget`
+    are also accepted and will be passed to the ``ProgressBar``.
+
+    Examples
+    --------
+
+    >>> def long_running(steps=10, delay=0.1):
+    ...     for i in progress(range(steps)):
```

Comment:
this will probably be such a common idiom `progress(range(steps))` ... we could provide a `prange` (or perhaps `progrange`, to avoid conflict with numba's prange?) convenience, similar to `tqdm.trange`:

```python
def progrange(*args, **kwargs):
    return progress(range(*args), **kwargs)

Reply:
:shrug: I also wouldn't mind leaving this for future deliberation. I like seeing `progress(range())` — very explicit.

Reply:
That’s fine too

Reply:
I've added in for now, if people think we should leave it out, we can. Personally I don't see the dangers of having it there - if people prefer to be explicit, they still can be. But happy to default to consensus